indispensable/src/commonMain/kotlin/at/asitplus/signum/indispensable/asn1/encoding/NumberEncoding.kt

```diff
@@ -241,222 +163,374 @@ fun Int.toUnsignedByteArray() = toLong().toUnsignedByteArray()
+ * while the highest bit indicates if more bytes are to come. Trailing bytes are ignored.
+ *
+ * @return the decoded unsigned BigInteger and the underlying varint-encoded bytes as `ByteArray`
+ */
+fun ByteArray.decodeAsn1VarBigInt(): Pair<BigInteger, ByteArray> = this.throughBuffer { it.decodeAsn1VarBigInt() }
+
+/**
+ * Decodes an UInt from bytes using varint encoding as used within ASN.1: groups of seven bits are encoded into a byte,
+ * while the highest bit indicates if more bytes are to come. Trailing bytes are ignored.
+ *
+ * @return the decoded UInt and the underlying varint-encoded bytes as `ByteArray`
+ * @throws IllegalArgumentException if the number is larger than [UInt.MAX_VALUE]
+ */
+@Throws(IllegalArgumentException::class)
+fun ByteArray.decodeAsn1VarUInt(): Pair<UInt, ByteArray> = this.throughBuffer { it.decodeAsn1VarUInt() }
+
+/**
+ * Converts this UUID to a BigInteger representation
+ */
+@OptIn(ExperimentalUuidApi::class)
+fun Uuid.toBigInteger(): BigInteger = BigInteger.fromByteArray(toByteArray(), Sign.POSITIVE)
+
+/**
+ * Tries to convert a BigInteger to a UUID. Only guaranteed to work with BigIntegers that contain the unsigned (positive)
+ * integer representation of a UUID, chances are high, though, that it works with random positive BigIntegers between
+ * 16 and 14 bytes large.
+ *
+ * Returns `null` if conversion fails. Never throws.
+ */
+@OptIn(ExperimentalUuidApi::class)
+fun Uuid.Companion.fromBigintOrNull(bigInteger: BigInteger): Uuid? =
+    catching { fromByteArray(bigInteger.toByteArray().ensureSize(16)) }.getOrNull()
+
+
+///////////KTX-IO
+
+/**
+ * Encodes this number using varint encoding as used within ASN.1: groups of seven bits are encoded into a byte,
+ * while the highest bit indicates if more bytes are to come
+ *
+ * @return the number of bytes written to the sink
+ */
+fun Sink.writeAsn1VarInt(number: ULong) = writeAsn1VarInt(number, ULong.SIZE_BITS)
 
 /**
  * Encodes this number using varint encoding as used within ASN.1: groups of seven bits are encoded into a byte,
  * while the highest bit indicates if more bytes are to come
+ *
+ * @return the number of bytes written to the sink
  */
-fun BigInteger.toAsn1VarInt(): ByteArray {
-    if (isZero()) return byteArrayOf(0)
-    require(isPositive) { "Only positive Numbers are supported" }
-    if (this < 128) return byteArrayOf(this.byteValue(exactRequired = true)) //Fast case
+fun Sink.writeAsn1VarInt(number: UInt) = writeAsn1VarInt(number.toULong(), UInt.SIZE_BITS)
+
+/**
+ * Encodes this number using varint encoding as used within ASN.1: groups of seven bits are encoded into a byte,
+ * while the highest bit indicates if more bytes are to come
+ *
+ * @return the number of bytes written to the sink
+ */
+private fun Sink.writeAsn1VarInt(number: ULong, bits: Int):Int {
+    if (number < 128u) return writeByte(number.toByte()).run { 1 } //Fast case
     var offset = 0
     var result = mutableListOf<Byte>()
```

Comment:
there has to be a way to refactor this and avoid the list

Reply:
be my guesst

Reply:
Unless overlooked something due to borked Github rendering of large PRs with many comments, this is the only outstanding issue and i am not the right person to wrap their head around this

Reply:
copy from `Sink.writeAsn1VarInt(number: BigInteger)`, with `bitLength` replacement as discussed